codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/HttpBoundProtocolPayloadGenerator.kt

```diff
@@ -0,0 +1,262 @@
+
+    private fun RustWriter.serializeViaPayload(
+        payloadMetadata: ProtocolPayloadGenerator.PayloadMetadata,
+        self: String,
+        member: MemberShape,
+        serializerGenerator: StructuredDataSerializerGenerator
+    ) {
+        val fnName = "serialize_payload_${member.container.name.toSnakeCase()}"
+        val ref = if (payloadMetadata.takesOwnership) "" else "&"
+        val serializer = RuntimeType.forInlineFun(fnName, operationSerModule) {
+            val outputT = if (member.isStreaming(model)) "ByteStream" else "ByteSlab"
+            it.rustBlockTemplate(
+                "pub fn $fnName(payload: $ref#{Member}) -> Result<#{$outputT}, #{BuildError}>",
+                "Member" to symbolProvider.toSymbol(member),
+                *codegenScope
+            ) {
+                val asRef = if (payloadMetadata.takesOwnership) "" else ".as_ref()"
+
+                if (symbolProvider.toSymbol(member).isOptional()) {
+                    withBlockTemplate(
+                        """
+                            let payload = match payload$asRef {
+                                Some(t) => t,
+                                None => return Ok(
+                            """,
+                     ")};",
+                        *codegenScope
+                    ) {
+                        when (val targetShape = model.expectShape(member.target)) {
+                            is StringShape, is BlobShape, is DocumentShape -> rust(
+                                // TODO Can we get rid of this allocation and make the function return a slice of bytes?
+                                """
+                                b"".to_vec()
+                                """
+                            )
+                            // If this targets a member and the member is `None`, return an "empty" `Vec<u8>`.
+                            is StructureShape -> rust("#T()", serializerGenerator.unsetStructure(targetShape))
+                        }
+                    }
+                }
+
+                withBlock("Ok(", ")") {
+                    renderPayload(member, "payload", serializerGenerator)
+                }
+            }
+        }
+        rust("#T($ref $self.${symbolProvider.toMemberName(member)})?", serializer)
+    }
+
+    private fun RustWriter.renderPayload(
+        member: MemberShape,
+        payloadName: String,
+        serializer: StructuredDataSerializerGenerator
+    ) {
+        when (val targetShape = model.expectShape(member.target)) {
+            is StringShape -> {
+                // Write the raw string to the payload.
+                // TODO Can we get rid of this allocation and make the function return a slice of bytes?
+                if (targetShape.hasTrait<EnumTrait>()) {
+                    rust("$payloadName.as_str().as_bytes().to_vec()")
+                } else {
+                    // TODO Cannot use `into_bytes()` because payload is behind shared reference.
```

Comment:
Should we have make these todos issues and reference them?

Reply:
I've fixed these. The answer to whether we can is yes for these particular serializer functions, but it's suboptimal because we then need to take ownership to put the data in the body, where we'd have to clone. Don't know what I was thinking when I wrote these.